### ISSUE TITLE
Various questions about 0.8.5

### ISSUE THREAD

**Issue by aminomancer** [MrOtherGuy/fx-autoconfig#38]
I've just been testing 0.8.5 and I have a few questions and suggestions. Just wanted to let you know nothing seems to be broken for me. I had to change a couple things since the API has changed slightly, but everything is working. The overall organization is really great too.

The new CSS features are pretty cool. I'd like to replace my ad hoc implementation with the new API. But I have some special considerations, particularly [because of this](https://github.com/aminomancer/uc.css.js#styling-browser-toolbox-windows). In order to get all this stuff (both stylesheets and scripts) to load in arbitrary browser toolboxes, I had to make config.js load the manifest from the root profile directory (so like `chrome_debugger_profile/../chrome/utils/chrome.manifest`), then make everything else that uses nsIFile navigate up to the root directory too.

This is a pretty specific use case, so I don't expect fx-autoconfig to explicitly support it. It's hard to support it as an opt-in setting, because the autoconfig loader itself needs to know where to look for the manifest very early in startup. So it's probably always gonna require modifying at least one file. But I suppose those instructions could be simplified a lot if all fx-autoconfig's file operations used the chrome registry. Then, within the browser toolbox context, we'd only need to do 1 thing differently — change the path of the manifest from `currentProfile/utils/chrome.manifest` to `rootProfile/utils/chrome.manifest`. And everything else could just follow from that, since the registered URLs would be the same in every context.

Once you've modified config.js so it can find the real manifest, this:

```js
Cc["@mozilla.org/chrome/chrome-registry;1"].getService(Ci.nsIChromeRegistry)
  .convertChromeURL(Services.io.newURI("chrome://userchrome/content/"))
  .QueryInterface(Ci.nsIFileURL).file.parent.path
```

already returns `path/to/root_profile/chrome/resources/` in the toolbox (rather than `path/to/chrome_debugger_profile/chrome/resources/`, which doesn't exist).

And that already seems to be the case with most of the file system functions now, like `FS.getStyleDir()`. But there are a few things that still get an nsIFile object by navigating from `UChrm`, which leads to the `chrome_debugger_profile` in this case. Specifically, `FileSystem.BASE_FILEURI`, `FileSystem.#getEntry`, and `FileSystem.chromeDir`. Idk how easy or desirable it is to refactor these to start from chrome URIs, since they are ultimately trying to get to the chrome dir, which is not itself registered.

But I think you could simply assume the chrome dir is the parent of one of the directories that _is_ registered, probably the utils folder. `chrome.manifest` already has to stipulate the relationship between itself and the chrome folder, e.g. `content userscripts ../JS/` only works if the manifest is in `profile/chrome/utils/`. So I think it wouldn't be unreasonable to just universally treat the chrome dir as the parent of the utils dir, since if it somehow is not, all of this breaks down anyway unless the user has made major modifications to all the major files.

Having done that, the starting point for the whole fx-autoconfig file system is then a function of the manifest location, which is stipulated by config.js. So you'd only have to modify one thing to make it dynamically locate a single installation of fx-autoconfig. If you do end up implementing INI configuration feature, you could just add a setting for that, instead of me modifying config.js directly.

---

Something that's impeding my adoption of the new CSS features is that the way they're injected precludes easily saving them from the style editor. When they are loaded via `file://` URI, you can live edit them. It has some downsides because there are some `:-moz-*` prefixed pseudo classes that check the principal and basically require the stylesheet to be registered. I have worked around that by confining those kinds of rules to [a separate file](https://github.com/aminomancer/uc.css.js/blob/master/resources/layout/uc-nsRules.au.css) that I won't need to edit as often.

A related issue is importing shared stylesheets. If the stylesheet is loaded from a chrome URI, then anything it imports needs to also be imported from a chrome URL. I'm currently doing `@import url(uc-globals.css)` from my agent sheet, which works since they're both in the same folder. Moving the agent sheet into the CSS folder, I could do `@import url(../uc-globals.css)`, but only if I loaded the agent sheet from its file URI. So as it is, I'd need to register uc-globals.css with a chrome URL, making it so neither can be saved from the devtools.

But it seems like all of this could still work using file URIs. So, there could be a metadata directive controlling whether a file gets injected with a chrome URL or a file URL.

```js
let uri = style.chromeURI;
if (style.useFileURI) {
  uri = FS.convertChromeURIToFileURI(uri);
}
sss.loadAndRegisterSheet(uri, sss.AGENT_SHEET);
ScriptData.markScriptRunning(style);
```

---

Another minor weakness of my approach to styling browser toolbox windows arises because fx-autoconfig uses prefs to manage disabling scripts. This might be resolved by your suggestion in #37 though. Basically, the toolbox window doesn't know which scripts are disabled because the toolbox profile may have its own prefs, until it's deleted or something. With INI configuration and the registry-based file navigation I described, the toolbox instance of fx-autoconfig would be able to find the root profile's config.

---

This could just be an oversight on my part, but I don't understand what's preventing agent sheets from trying (and failing) to load twice. You're adding them to the `agentStyleSet`, which gets loaded by the stylesheet service. But then you're also adding them to `this.styles`, like the author sheets. Then on DOMContentLoaded, you check `aScript.type === "style" && aScript.styleSheetMode === "author"` and return early for author sheets. But agent sheets should just carry on and ultimately reach this:

```js
const injection = aScript.isESM
  ? ScriptData.injectESMIntoGlobal(aScript,win)
  : ScriptData.injectClassicScriptIntoGlobal(aScript,win);
```

So isn't it going to try executing an agent sheet as JS with the subscript loader? I might be missing something. I didn't notice any errors logged, but idk whether to expect any in this case.

---

I noticed you're using `ChromeUtils.compileScript` now for ESM window/page scripts. So I figured this could have some performance benefits. But I'm curious what's the reasoning behind the data URI?

```js
ChromeUtils.compileScript(`data:,"use strict";import("${aScript.chromeURI.spec}").catch(console.error)`)
```

Is there any reason to think that could change behavior or perhaps how the debugger interacts with scripts?

**Comment by MrOtherGuy** [comment]
> I had to make config.js load the manifest from the root profile directory (so like chrome_debugger_profile/../chrome/utils/chrome.manifest), then make everything else that uses nsIFile navigate up to the root directory too.

Is there a reason you edit config.js here? I kinda do the same thing in the [test_profile](https://github.com/MrOtherGuy/fx-autoconfig/blob/6127fb08e98132c4393cf3cc4fce597be2b35c0e/test_profile/chrome/utils/chrome.manifest) where the manifest maps chrome paths to point to other locations. Any such re-mappings *should* happen through the manifest and I don't immediately see why it wouldn't work for the browser_toolbox profile.

> So I think it wouldn't be unreasonable to just universally treat the chrome dir as the parent of the utils dir, since if it somehow is not, all of this breaks down anyway unless the user has made major modifications to all the major files.

So, the basis of file system operations would be the **parent** of whatever directory `chrome://userchrome/content/` resolves to and not `UChrm`? This sounds wrong to me to be honest. In my view, the whole operating context of fx-autoconfig should be the active profile and by extension *its* chrome directory. See, in test_profile I have this structure:

```
<profile>
  |__chrome
  | |__css
  | |__utils
  |   |__boot.sys.mjs
  |   |__chrome.manifest
  |__test_profile
    |__chrome
      |__tests
      |__utils
         |__chrome.manifest
```

The chrome.manifest of the `test_profile` is what config.js ends up loading. That manifest maps:

* `utils` of the parent as `chrome://userchromejs/content/`
* `tests` as `chrome://userscripts/content/`
* `css` as `chrome://userstykes/skin/`

So here, the loader and styles exist outside of the current profile. But indeed, like you have said, FS operations still try to access file *inside the active profile*. This is arguably wrong, `FS.getEntry()` does this as a hacky way to support relative url (like `../`), and simply appends it to `UChrm`. That is... well it's not great to say the least, but it would be more wrong to always get the parent of `utils` and appending the path to that. In any event, I should definitely change `#getEntry` to not look for `UChrm` specifically, but to one of defined folders instead.

> But it seems like all of this could still work using file URIs. So, there could be a metadata directive controlling whether a file gets injected with a chrome URL or a file URL.

Yep, I could totally do that, this is great suggestion, thanks! The main reason for having them be `chrome://` urls is exactly what you pointed out - some features are not available for `file://` uri style sheets. Calling `_ucUtils.updateStyleheet(sheetname,"author")` should update the style though - of course, it's still not really live-editable, but it's something.

> This could just be an oversight on my part, but I don't understand what's preventing agent sheets from trying (and failing) to load twice. You're adding them to the agentStyleSet, which gets loaded by the stylesheet service. But then you're also adding them to this.styles, like the author sheets.

Okay, the `agentStyleSet` is not to prevent duplicate entries. It is just to store a *subset* of styles that are agent styles. And then after all style files are read by the loader send that subset to StyleSheetService to be registered all at once. During `onDOMContent` they should end up and [return here](https://github.com/MrOtherGuy/fx-autoconfig/blob/6127fb08e98132c4393cf3cc4fce597be2b35c0e/profile/chrome/utils/boot.sys.mjs#L174) because any `Style` scripts are also marked with `noExec = true` during parsing. The reason for `styles` and `scripts` arrays is to store objects that the loader is aware of - regardless of if they are enabled or not. Functionally we could just forget agent styles after parsing because we don't do anything about them afterwards (except generate menubar menu items) but its easier to keep reference to them for the whole runtime.

> I noticed you're using ChromeUtils.compileScript now for ESM window/page scripts. So I figured this could have some performance benefits. But I'm curious what's the reasoning behind the data URI?

Honestly, please do something about it :)
No, really this was the only way I found with which I was able to make the script file itself a module script. I mean here, the data-uri "script" is not loaded as module script, it is just a wrapper that loads the script you care about as a module script. If only `Services.scriptloader.loadSubScriptWithOptions()` would have an option to inject the subscript as module, but alas it doesn't.

**Comment by MrOtherGuy** [comment]
I've added support for a new `@usefileuri` meta-tag in 65fac03 - actually the patch also removes that extra `agentStyleSet` construction and instead just filters them out during call to `addAgentStyles()` - this should make the intention more clear, hopefully.

**Comment by aminomancer** [comment]
> > I had to make config.js load the manifest from the root profile directory (so like chrome_debugger_profile/../chrome/utils/chrome.manifest), then make everything else that uses nsIFile navigate up to the root directory too.
> 
> Is there a reason you edit config.js here? I kinda do the same thing in the [test_profile](https://github.com/MrOtherGuy/fx-autoconfig/blob/6127fb08e98132c4393cf3cc4fce597be2b35c0e/test_profile/chrome/utils/chrome.manifest) where the manifest maps chrome paths to point to other locations. Any such re-mappings _should_ happen through the manifest and I don't immediately see why it wouldn't work for the browser_toolbox profile.

The browser toolbox profile doesn't have a manifest at all. These profiles are created on the spot when opening the toolbox, and I delete them pretty frequently. Plus, a toolbox profile can have an arbitrary number of additional toolbox profiles nested within each other, if you open a toolbox on another toolbox. So that would require you to recreate files in every profile, every time a new one is created. Editing config.js is something I did once several years ago and I haven't needed to touch it again (until this recent change to .sys.mjs). That time saving is the rationale behind all these changes.

You're right that adding a custom manifest for the toolbox profile is less effort than just duplicating the whole chrome directory in the toolbox profile. But `profile/chrome_debugger_profile` needs its own paths, and `profile/chrome_debugger_profile/chrome_debugger_profile` needs its own, and so on. I guess there could be ways of automating it, but it's so much easier to just tell config.js to look for the manifest in the real profile. So far, I've never had any cause to put custom files in the chrome_debugger_profile. Like, when I want to customize the toolbox specifically, I wouldn't give it its own chrome folder, I would just add a script to my real profile and restrict it to the toolbox URL.

I get what you're saying, fx-autoconfig is basically installed to a profile, and all its filesystem operations should be relative to that profile. I guess the way I'm thinking about this, fx-autoconfig would be installed to a "real" profile, and then automagically apply to every "child" toolbox profile. Of course, if someone actually installed fx-autoconfig in the `chrome_debugger_profile/chrome` folder, then it could load that first, and only try to load the "real" profile's files as a fallback, when a manifest is not found within `chrome_debugger_profile`.

> > This could just be an oversight on my part, but I don't understand what's preventing agent sheets from trying (and failing) to load twice. You're adding them to the agentStyleSet, which gets loaded by the stylesheet service. But then you're also adding them to this.styles, like the author sheets.
> 
> any `Style` scripts are also marked with `noExec = true` during parsing.

Ah man, I knew I must have been missing something obvious like that. My bad.

> The reason for `styles` and `scripts` arrays is to store objects that the loader is aware of - regardless of if they are enabled or not. Functionally we could just forget agent styles after parsing because we don't do anything about them afterwards (except generate menubar menu items) but its easier to keep reference to them for the whole runtime.

I see what you mean. That makes sense now!


> > I noticed you're using ChromeUtils.compileScript now for ESM window/page scripts. So I figured this could have some performance benefits. But I'm curious what's the reasoning behind the data URI?
> 
> Honestly, please do something about it :) No, really this was the only way I found with which I was able to make the script file itself a module script. I mean here, the data-uri "script" is not loaded as module script, it is just a wrapper that loads the script you care about as a module script. If only `Services.scriptloader.loadSubScriptWithOptions()` would have an option to inject the subscript as module, but alas it doesn't.

That's fascinating. If I understand correctly, the goal here is to load an ESM page script in the chrome window in the same way we can load .mjs files in content. The motivation being to give the script access to static `import` declarations? Actually, I would have expected `<script type="module" src="${aScript.chromeURI.spec}">` to work out of the box. But I guess the chrome window is pretty weird. I'll try messing around with it sometime.

---

> I've added support for a new `@usefileuri` meta-tag in [65fac03](https://github.com/MrOtherGuy/fx-autoconfig/commit/65fac03c4da8d06b6e7e89a1e235dedfd9397719) - actually the patch also removes that extra `agentStyleSet` construction and instead just filters them out during call to `addAgentStyles()` - this should make the intention more clear, hopefully.

Awesome! Yeah, removing the set helps clarify the part I got confused about.

**Comment by aminomancer** [comment]
By the way, one other suggestion that just occurred to me — I mentioned earlier how I have 2 author sheets. Till now, the main one was loaded by file URI, and it imported the second one by chrome URI. So the cascading order was explicit that way.

With the new `@usefileuri` tag, I can remove the CSS import and just put both author sheets in the CSS folder, with the main one using the new tag, and the second one not using it. But the cascading order then becomes dependent on file names (I think). So I went to check if I can use `@loadOrder`, but it looks like that only works for scripts. Not a big deal for my use case since I can just keep using CSS import for now, but it could be a nice feature for 0.9.

**Comment by MrOtherGuy** [comment]
> Plus, a toolbox profile can have an arbitrary number of additional toolbox profiles nested within each other, if you open a toolbox on another toolbox. So that would require you to recreate files in every profile, every time a new one is created. Editing config.js is something I did once several years ago and I haven't needed to touch it again (until this recent change to .sys.mjs). That time saving is the rationale behind all these changes.

Fair enough. I feel like an argument could be made that chrome_debugger_profile *should* behave like you describe - so that it inherits the autoconfig from the parent, but I think there is a bit of a risk there with somebody then accidentally affecting the toolbox profile causing who-knows-what. Besides, if editing config.js is all you need for your purpose then I suppose that's fine.
Another possibility that comes to mind, I suppose, is to have a script generate a modified copy of the chrome.manifest when toolbox profile is created.

> That's fascinating. If I understand correctly, the goal here is to load an ESM page script in the chrome window in the same way we can load .mjs files in content. The motivation being to give the script access to static import declarations? Actually, I would have expected <script type="module" src="${aScript.chromeURI.spec}"> to work out of the box. But I guess the chrome window is pretty weird. I'll try messing around with it sometime.

Pretty much, yeah. In addition to static import declarations, a benefit of module scripts is that their variables exist in their own scope. So if in *normal* script you had a global function or variable it would also be available for all scripts in the window - so you need to be careful not to use same file symbols in your scripts. I mean, you can of course enclose them to immediately invoked functions like `(function(){ let blaa = 2 })()` or scope them in some other manner. But with module scripts they just live their own "global" scope and won't interfere with each other.

Man I can tell you, I tried a lot of different things to make the scripts load as modules. I believe `<script type="module">` didn't work because it wouldn't actually execute. Certainly, using the data:uri wrapper feels wrong - but honestly it works quite well as far as I can tell.

As far as `@loadOrder` is concerned, it only doesn't work because I'm not sorting styles based on their declared load order, which I simply because I didn't think it makes any difference for scripts. But you're right, it *does* matter for cascade. This is easily fixed by just sorting the loader styles array.

**Comment by MrOtherGuy** [comment]
Anyhow, 040f6fa39acd8397f81b7db001136778df3e42ba makes FileSystem operations convert relative urls correctly using chrome registry - I think that should work as a solution to the issue you had.

Styles get sorted using `@loadOrder` with be85166bdba6cded2b2be3f626bdee41d52610c2 and indeed, this makes it affect cascade order as you would expect.

I've also removed few "useless" properties of fs.sys.mjs that weren't really useful to use, including `BASE_FILEURI` which doesn't make sense because the question is "base of what?" when the three directories might not be in the same directory anyway. There's now `.RESOURCE_URI`, `SCRIPT_URI` and `STYLE_URI` instead all of which return a new `Ci.nsIFileURL` for their path converted using chrome registry. And also removed is `.PROFILE_DIR` which was in useless format anyway. Still, I've bumped version to `0.8.6` for these changes.

**Comment by aminomancer** [comment]
> Fair enough. I feel like an argument could be made that chrome_debugger_profile _should_ behave like you describe - so that it inherits the autoconfig from the parent, but I think there is a bit of a risk there with somebody then accidentally affecting the toolbox profile causing who-knows-what.

We're definitely agreed on that, it's a nonstandard usage and a bit of an edge case anyway. With your latest commit, it seems like I can accomplish this with a single edit to config.js now, so that's great.

> Besides, if editing config.js is all you need for your purpose then I suppose that's fine. Another possibility that comes to mind, I suppose, is to have a script generate a modified copy of the chrome.manifest when toolbox profile is created.

Yeah... that might be a better approach. The [profile creation](https://searchfox.org/mozilla-central/rev/4e22b886bbd4c274f268c4a285ab7da00e95c99b/devtools/client/framework/browser-toolbox/Launcher.sys.mjs#177) could just be intercepted. But a problem I've been running into a lot lately is the increasing use of private class fields/methods. UrlbarView was changed this year to almost entirely use private fields, which broke a ton of my mods all at once. I ended up needing to completely override the entire class.

It's very hard to maintain because you can't just override the file either, because it has a resource URI. Same with BrowserToolboxLauncher. Ideally, I would just copy the file from m-c when it changes, and replace all the hashes in keys with underscores or something. But since the class needs to be overrided from JS, the code has to be modified in a bunch of other ways too, so any update becomes really tedious.

I filed [a bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1801966) to try and provide some workaround, but it's not really practical. I just avoid using private fields at all now, but it seems unlikely the trend will stop, nor the existing uses be rolled back :(

> Pretty much, yeah. In addition to static import declarations, a benefit of module scripts is that their variables exist in their own scope. So if in _normal_ script you had a global function or variable it would also be available for all scripts in the window - so you need to be careful not to use same file symbols in your scripts. I mean, you can of course enclose them to immediately invoked functions like `(function(){ let blaa = 2 })()` or scope them in some other manner. But with module scripts they just live their own "global" scope and won't interfere with each other.

Ah, that's a great point. The extra indentation from wrapping everything in IIFEs is annoying.

> Man I can tell you, I tried a lot of different things to make the scripts load as modules. I believe `<script type="module">` didn't work because it wouldn't actually execute. Certainly, using the data:uri wrapper feels wrong - but honestly it works quite well as far as I can tell.

Yeah, if it ain't broke don't fix it. And I imagine there will probably be some effort at Mozilla to convert browser.js and related scripts to ES modules. Most of the scripts in browser/base/content/ really feel like they're supposed to be modules anyway. And now all new development on content seems to be done with mjs files, all the new widgets/components (storybook and lit element stuff) are mjs files as well. So that's probably not so far around the corner.

> Anyhow, [040f6fa](https://github.com/MrOtherGuy/fx-autoconfig/commit/040f6fa39acd8397f81b7db001136778df3e42ba) makes FileSystem operations convert relative urls correctly using chrome registry - I think that should work as a solution to the issue you had.
> 
> Styles get sorted using `@loadOrder` with [be85166](https://github.com/MrOtherGuy/fx-autoconfig/commit/be85166bdba6cded2b2be3f626bdee41d52610c2) and indeed, this makes it affect cascade order as you would expect.
> 
> I've also removed few "useless" properties of fs.sys.mjs that weren't really useful to use, including `BASE_FILEURI` which doesn't make sense because the question is "base of what?" when the three directories might not be in the same directory anyway. There's now `.RESOURCE_URI`, `SCRIPT_URI` and `STYLE_URI` instead all of which return a new `Ci.nsIFileURL` for their path converted using chrome registry. And also removed is `.PROFILE_DIR` which was in useless format anyway. Still, I've bumped version to `0.8.6` for these changes.

That's great! That all makes sense. Thanks for the new version, I really appreciate it.

---

At this point I'm thinking about converting my scripts to ESM, but it seems like I'll need to change the filename, which I think would disrupt the git history. I'm actually not sure if it will so I'll need to test that. Changing the filename will also interfere with my userscript manager's update procedure, so if anyone is actually using that, they wouldn't hear about the update. Anyway, I realized you have `@backgroundmodule` for system modules, but no meta tag for ESM window scripts. And if I'm not mistaken, the dynamic import can handle any arbitrary file extension. So it might also be worth adding a tag like `@esmodule` or something.

But I'm not sure, maybe the large number of ways to change how the script is processed is going to get confusing. It could be a good reason to have a tag with a value.

```js
// @sourceType   script|module|sys.mjs|jsm
```

I like the name `sourceType` since this reminds me of ESLint config's `parserOptions.sourceType`. But I can't think of a good name to differentiate system ESM from system JSM while still differentiating system ESM from window ESM. That distinction between system ESM and window ESM is also a potential spot for confusion.

I guess you could just abandon support for JSM, since it's not long for this world anyway. I imagine you were probably thinking of doing that at some point. And anyone who's currently using JSM but is updating fx-autoconfig could just convert to ESM, which will make their module neater anyway. Then the scheme could be `script|module|system` for parity with `uc.js|uc.mjs|sys.mjs`.

**Comment by MrOtherGuy** [comment]
> I guess you could just abandon support for JSM, since it's not long for this world anyway. I imagine you were probably thinking of doing that at some point. And anyone who's currently using JSM but is updating fx-autoconfig could just convert to ESM, which will make their module neater anyway. Then the scheme could be script|module|system for parity with uc.js|uc.mjs|sys.mjs.

Yes, I intend to deprecate `@backgroundmodule` at some point. I don't know when though, it kinda depends on how long `ChromeUtils.import()` which is used to load JSMs stays in the tree. In the mean time, it would sure be nice to figure out what would be a good way to differentiate between these different script "types". Currently it's really just been "Add a new method, add a new meta tag" kind of thing.

I believe there would indeed be 3 kinds of scripts like you said, I can't think of anything else really. Well, I suppose technically wasm could be possible - not sure why that would ever be a good idea in this context but still. Possibly worker scripts too - though I can't think of use case for that.

I gotta say, I kinda like the implicit filename based selection but sure, the kind of `@sourceType` would also work. But point being, I think it would maybe be best to support one or the other - not a mix of both. Related to that, I believe at some point the file extension did actually matter for `ChromeUtils.importESModule()` where it would treat `.sys.mjs` one way and `.jsm` another. So, obviously if the filename matters for some internal calls then we can't really do much about that.

**Comment by aminomancer** [comment]
> I gotta say, I kinda like the implicit filename based selection but sure, the kind of `@sourceType` would also work. But point being, I think it would maybe be best to support one or the other - not a mix of both. Related to that, I believe at some point the file extension did actually matter for `ChromeUtils.importESModule()` where it would treat `.sys.mjs` one way and `.jsm` another. So, obviously if the filename matters for some internal calls then we can't really do much about that.

Oh that's interesting, well fair enough, this is a pretty minor thing anyway. I think that's just about everything so feel free to close this. Thanks again!

**Comment by MrOtherGuy** [comment]
Yes, there's definitely still *something* going on with file names, because if you do for example `Cu.isModuleLoaded("chrome://userchromejs/content/boot.sys.mjs")` then that will return `false`. It will, however return `true` for `".../boot.jsm"`

I don't know exactly what this would mean, but there's totally something weird going on related to file names, so we need to be a bit careful in that regard.

**Comment by aminomancer** [comment]
Seems like this maybe?

https://searchfox.org/mozilla-central/rev/7042a4a444f040ed87045c216f97cb73ce9410df/js/xpconnect/loader/mozJSModuleLoader.cpp#1162

https://searchfox.org/mozilla-central/rev/7042a4a444f040ed87045c216f97cb73ce9410df/js/xpconnect/loader/mozJSModuleLoader.cpp#154

It looks like there might be an oversight, like it wasn't expected to be passed real ESMs.

**Comment by MrOtherGuy** [comment]
I'm not sure about oversight, kinda feels like maybe intentional - since new code should know to call isESModuleLoaded or isJSModuleLoaded it feels more like being a wrapper for some legacy code.

But more importantly, to me that is an indication that there is *something* fishy happening with file names and as such we may not be able to completely freely decide what file extensions to use. At least not until `.jsm` support is removed altogether.
